norse/torch/functional/stdp.py

```diff
@@ -1,400 +1,276 @@
-        stride (int or tuple): convolutional stride, defaults to 1 like nn.Conv2d
-        padding (int or tuple): convolutional padding, defaults to 0 like nn.Conv2d
-        dilation (int or tuple): convolutional dilation, defaults to 1 like nn.Conv2d
-    """
-
-    a_pre: torch.Tensor = torch.as_tensor(1.0)
-    a_post: torch.Tensor = torch.as_tensor(1.0)
-    tau_pre_inv: torch.Tensor = torch.as_tensor(1.0 / 100e-3)
-    tau_post_inv: torch.Tensor = torch.as_tensor(1.0 / 100e-3)
-    w_min: float = 0.0
-    w_max: float = 1.0
-    eta_plus: float = 1.0
-    eta_minus: float = 1.0
-    stride: Union[int, Tuple] = 1
-    padding: Union[int, Tuple] = 0
-    dilation: Union[int, Tuple] = 1
-
-
-class STDPState(NamedTuple):
+# STDP state class
+class STDPState:
     """State of spike-timing-dependent plasticity (STDP)
-
-    Parameters:
-        x (torch.Tensor): presynaptic spike trace
-        y (torch.Tensor): postsynaptic spike trace
-    """
-
-    x: torch.Tensor
-    y: torch.Tensor
-
-
-def _A_plus_soft(w, w_max, eta_plus):
-    """Soft upper bound for multiplicative STDP
-
     Parameters:
-        w (torch.Tensor): synaptic weight
-        w_max (float): upper bound on synaptic weight (should be > w_min)
-        eta_plus (float): contribution of pre-post activity to synaptic weight (should be >= 0)
+        t_pre (torch.Tensor): presynaptic spike trace
+        t_post (torch.Tensor): postsynaptic spike trace
     """
-    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
-    return torch.nn.functional.relu(w_max - w) * eta_plus
 
+    def __init__(self, t_pre, t_post):
+        self.t_pre = t_pre
+        self.t_post = t_post
 
-def _A_plus_hard(w, w_max, eta_plus):
-    """Hard upper bound for multiplicative STDP
-
-    Parameters:
-        w (torch.Tensor): synaptic weight
-        w_max (float): upper bound on synaptic weight (should be > w_min)
-        eta_plus (float): contribution of pre-post activity to synaptic weight (should be >= 0)
-    """
-    return heaviside(w_max - w) * eta_plus
-
-
-def _A_minus_soft(w, w_min, eta_minus):
-    """Soft lower bound for multiplicative STDP
-
-    Parameters:
-        w (torch.Tensor): synaptic weight
-        w_min (float): lower bound on synaptic weight (should be < w_max)
-        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
-    """
-    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
-    return torch.nn.functional.relu(w - w_min) * eta_minus
-
-
-def _A_minus_hard(w, w_min, eta_minus):
-    """Hard lower bound for multiplicative STDP
-
+    def decay(
+        self,
+        z_pre,
+        z_post,
+        tau_pre_inv,
+        tau_post_inv,
+        a_pre,
+        a_post,
+        dt,
+    ):
+        self.t_pre += dt * tau_pre_inv * (-self.t_pre + a_pre * z_pre)
+        self.t_post += dt * tau_post_inv * (-self.t_post + a_post * z_post)
```

Comment:
Don't in-place ops destroy gradient flow?

Reply:
Just checked. They do. Thanks for the heads up.